src/Data/Text/Utf8/Unlower.hs

```diff
@@ -25,23 +25,37 @@ import qualified Data.List as List
 --    unlowerCodePoint 'ß' == "ẞß"
 --
 unlowerCodePoint :: Char -> [Char]
-unlowerCodePoint c =
-  maybe [c] id $ HashMap.lookup c unlowerings
-
-
--- | Reference implementation of unlowerCodePoint, simple but slow.
--- This is useful fo
-refUnlowerCodePoint :: Char -> [Char]
-refUnlowerCodePoint =
+unlowerCodePoint =
   \c -> maybe [] id $ HashMap.lookup c completeMapping
   where
     addUnlowering c hm =
       HashMap.insertWith (++) (Char.toLower c) [c] hm
+    -- This mapping becomes a top level definition and is only evaluated once.
```

Comment:
This seems far to performance critical to rely on compiler optimizations to fire just right.
I'd instead introduce an explicit top-level definition that is marked `NOINLINE`.

Reply:
Well It's good that you mentioned it, because I forgot to remove the trivial mappings so the HashMap contained 1114112 entries. I changed it to an IntMap while I was at it, and it has 2771 entries where almost half are pointing to the empty list (because the characters are upper case).